Dockerfile

```diff
@@ -1,26 +1,90 @@
-from alpine as build-environment
+# syntax=docker/dockerfile:1.4
+FROM alpine:3.15 AS build-environment
 WORKDIR /opt
-RUN apk add clang lld curl build-base linux-headers git \
-    && curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs > rustup.sh \
-    && chmod +x ./rustup.sh \
-    && ./rustup.sh -y
+
+RUN apk update
+
 WORKDIR /opt/foundry
+
+RUN set -eux; \
+	\
+	apk add --no-cache --virtual .foundry-deps \
+		ca-certificates \
+		clang \
+		lld \
+        build-base \
+        linux-headers \
+        git \
+        curl \
+        findutils \
+	; \
+	\
+	dpkgArch="$(dpkg --print-architecture | awk -F- '{ print $NF }')"; \
+	curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs > rustup.sh; \
+    chmod +x ./rustup.sh; \
+    ./rustup.sh -y; \
+    export CARGO_HOME=/opt/foundry/.cargo;
+
+ENV CARGO_HOME=/opt/foundry/.cargo
+
+# Copy seperate
+COPY ./Cargo.toml /opt/foundry/Cargo.toml
+COPY ./Cargo.lock /opt/foundry/Cargo.lock
 COPY . .
-RUN source $HOME/.profile && cargo build --release \
-    && strip /opt/foundry/target/release/forge \
-    && strip /opt/foundry/target/release/cast \
-    && strip /opt/foundry/target/release/anvil
 
-from alpine as foundry-client
+RUN source $HOME/.profile && cargo build --release
+
+# build for release
+RUN rm -rf /opt/foundry/target/release/deps/
+RUN rm -rf /opt/foundry/target/release/build/
+
+# Strip binaries
+RUN find /opt/foundry/target/release -type f -exec file {} + | awk -F: '/:.*ELF/{print $1}' | xargs strip --strip-all
+RUN find /opt/foundry/target/release -type f | xargs file | grep 'ELF.*stripped' | cut -f1 -d: | xargs strip --strip-all
+
+# Cleanup deps
+RUN apk del --no-network .foundry-deps;
+
+FROM alpine:3.15 AS foundry-client
+
+# TODO: Explore using `docker.io/frolvlad/alpine-glibc:alpine-3.15_glibc-2.34`
+
 ENV GLIBC_KEY=https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub
 ENV GLIBC_KEY_FILE=/etc/apk/keys/sgerrand.rsa.pub
 ENV GLIBC_RELEASE=https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.35-r0/glibc-2.35-r0.apk
 
-RUN apk add linux-headers gcompat
-RUN wget -q -O ${GLIBC_KEY_FILE} ${GLIBC_KEY} \
-    && wget -O glibc.apk ${GLIBC_RELEASE} \
-    && apk add glibc.apk --force
+RUN set -eux; \
+	\
+	apk add --no-cache --virtual .musl-deps \
+        linux-headers \
+        gcompat \
+	; \
+	\
+    wget -q -O ${GLIBC_KEY_FILE} ${GLIBC_KEY}; \
+    wget -O glibc.apk ${GLIBC_RELEASE}; \
+    apk add glibc.apk --force; \
+    apk del --no-network .musl-deps;
+
 COPY --from=build-environment /opt/foundry/target/release/forge /usr/local/bin/forge
 COPY --from=build-environment /opt/foundry/target/release/cast /usr/local/bin/cast
 COPY --from=build-environment /opt/foundry/target/release/anvil /usr/local/bin/anvil
-ENTRYPOINT ["/bin/sh", "-c"]
\ No newline at end of file
+
+EXPOSE 8545/tcp
+EXPOSE 8545/udp
+EXPOSE 8180
+EXPOSE 3001/tcp
+
+STOPSIGNAL SIGQUIT
+
+ENTRYPOINT ["/bin/sh", "-c"]
```

Comment:
Not entirely sure on the difference between `ENTRYPOINT` and `CMD` but there is another PR talking about that (#2196). Thoughts on that?